crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -2312,6 +2346,116 @@ unsafe impl<T: ?Sized> QueryData for PhantomData<T> {
 /// SAFETY: `PhantomData` never accesses any world data.
 unsafe impl<T: ?Sized> ReadOnlyQueryData for PhantomData<T> {}
 
+/// A compile-time checked union of two different types that differs based on the
+/// [`StorageType`] of a given component.
+pub(super) union StorageSwitch<C: Component, T: Copy, S: Copy> {
+    table: ManuallyDrop<T>,
```

Comment:
I don't think you need to wrap these in `ManuallyDrop` if you're also constraining `T` and `S` to be `Copy`.